### PR TITLE
feat: handle 'session not found' errors

### PR DESCRIPTION
**Pull request by oprudkyi**
fixes https://github.com/colopl/laravel-spanner/issues/37

There are a few cases when a 'Session not found' error can
[happen](https://cloud.google.com/spanner/docs/sessions#handle_deleted_sessions):

 - Scripts that idle too long - for example, a [Laravel queue worker](https://laravel.com/docs/9.x/queues#the-queue-work-command)
or anything that doesn't call Spanner frequently enough (more than once an hour).
 - The session is more than 28 days old.
 - Some random flukes on Google's side.

### CODE REVIEW

File: src/Colopl/Spanner/Concerns/ManagesSessionPool.php [hunk 1]
```diff
@@ -81,7 +81,8 @@ public function listSessions(): Collection
     {
         $databaseName = $this->getSpannerDatabase()->name();
         $response = (new ProtobufSpannerClient())->listSessions($databaseName);
-        return collect($response->iterateAllElements())->map(function (ProtobufSpannerSession $session) {
+        return collect($response->iterateAllElements())->map(function ($session) {
+            assert($session instanceof ProtobufSpannerSession);
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Concerns/ManagesSessionPool.php:85]: Why was this changed to an assertion?

File: src/Colopl/Spanner/Concerns/ManagesStaleReads.php [hunk 2]
```diff
@@ -56,7 +56,9 @@ public function cursorWithTimestampBound($query, $bindings = [], TimestampBoundI
      */
     public function selectWithTimestampBound($query, $bindings = [], TimestampBoundInterface $timestampBound = null): array
     {
-        return iterator_to_array($this->cursorWithTimestampBound($query, $bindings, $timestampBound));
+        return $this->sessionNotFoundWrapper(function () use ($query, $bindings, $timestampBound) {
+			return iterator_to_array($this->cursorWithTimestampBound($query, $bindings, $timestampBound));
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Concerns/ManagesStaleReads.php:60]: Please use 4 spaces instead of tabs.

File: src/Colopl/Spanner/Concerns/ManagesStaleReads.php [hunk 3]
```diff
@@ -68,7 +70,11 @@ public function selectWithTimestampBound($query, $bindings = [], TimestampBoundI
      */
     public function selectOneWithTimestampBound($query, $bindings = [], TimestampBoundInterface $timestampBound = null): ?array
     {
-        return $this->cursorWithTimestampBound($query, $bindings, $timestampBound)->current();
+        $result = $this->sessionNotFoundWrapper(function () use ($query, $bindings, $timestampBound) {
+			return $this->cursorWithTimestampBound($query, $bindings, $timestampBound)->current();
+		});
+        assert(is_null($result) || is_array($result));
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Concerns/ManagesStaleReads.php:76]: Why is there an assertion here?

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php [hunk 4]
```diff
@@ -158,7 +161,14 @@ protected function performRollBack($toLevel)
 
         if ($this->currentTransaction !== null) {
             if ($this->currentTransaction->state() === Transaction::STATE_ACTIVE) {
-                $this->currentTransaction->rollBack();
+                try {
+                    $this->currentTransaction->rollBack();
+                } catch (NotFoundException $e) {
+                    // ignore session not found error
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Concerns/ManagesTransactions.php:167]: Wouldn't ignoring the error cause unexpected behaviour upstream?

File: src/Colopl/Spanner/Connection.php [hunk 5]
```diff
@@ -443,4 +462,83 @@ protected function runQueryCallback($query, $bindings, Closure $callback)
...
+            throw new InvalidArgumentException("Unsupported sessionNotFoundErrorMode [{$handlerMode}].");
+        }
+        try {
+            return $callback();
+        } catch (NotFoundException $e) {
+            // ensure if this really error with session
+            if ($this->causedBySessionNotFound($e)) {
+                if ($this->inTransaction()) {
+                    // if we inside transaction then throw abort exception
+                    throw new AbortedException(self::SESSION_NOT_FOUND_CONDITION);
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:506]: This removes all context information given by `$e`. I think `$e` should be passed to AbortedException.

File: src/Colopl/Spanner/Connection.php [hunk 6]
```diff
@@ -443,4 +462,83 @@ protected function runQueryCallback($query, $bindings, Closure $callback)
...
+        }
+    }
+
+    /**
+     * Check if this is "session not found" error
+     *
+     * @param  Throwable  $e
+     * @return boolean
+     */
+    public function causedBySessionNotFound($e)
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:538]:
```suggestion
    public function causedBySessionNotFound(Throwable $e): bool
```

File: src/Colopl/Spanner/Connection.php [hunk 7]
```diff
@@ -82,6 +84,21 @@ class Connection extends BaseConnection
...
+
+    /**
+     * Try to maintain and then clear session pool on 'session not found' error
+     */
+    const CLEAR_SESSION_POOL = 'CLEAR_SESSION_POOL';
+
+    /**
+     * Used to detect specific exception
+     */
+    const SESSION_NOT_FOUND_CONDITION = 'Session does not exist';
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:100]:
```suggestion
    public const SESSION_NOT_FOUND_CONDITION = 'Session does not exist';
```

File: src/Colopl/Spanner/Connection.php [hunk 8]
```diff
@@ -82,6 +84,21 @@ class Connection extends BaseConnection
...
 
+    /**
+     * Try to maintain session pool on 'session not found' error
+     */
+    const MAINTAIN_SESSION_POOL = 'MAINTAIN_SESSION_POOL';
+
+    /**
+     * Try to maintain and then clear session pool on 'session not found' error
+     */
+    const CLEAR_SESSION_POOL = 'CLEAR_SESSION_POOL';
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:95]:
```suggestion
    public const CLEAR_SESSION_POOL = 'CLEAR_SESSION_POOL';
```

File: src/Colopl/Spanner/Connection.php [hunk 9]
```diff
@@ -82,6 +84,21 @@ class Connection extends BaseConnection
      */
     protected $sessionPool;
 
+    /**
+     * Try to maintain session pool on 'session not found' error
+     */
+    const MAINTAIN_SESSION_POOL = 'MAINTAIN_SESSION_POOL';
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:90]:
```suggestion
    public const MAINTAIN_SESSION_POOL = 'MAINTAIN_SESSION_POOL';
```

**Comment by taka-oyama** [comment]
Hi thanks for the pull request! I have some review feedback. Please check it out.

File: src/Colopl/Spanner/Concerns/ManagesSessionPool.php (hunk 1 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesSessionPool.php:85]: Hi @taka-oyama 
I got phpstan error there (probably because phpstan version not fixed in composer.json it installed the most recent version of phpstan - you probably will get the same errors if delete composer.lock and rerun tests/install) 
```
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Colopl/Spanner/Concerns/ManagesSessionPool.php (in context of class Colopl\Spanner\Connection)                                                            
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------- 
  84     Parameter #1 $callback of method Illuminate\Support\Collection<(int|string),mixed>::map() expects callable(mixed, (int|string)): Colopl\Spanner\Session,  
         Closure(Google\Cloud\Spanner\V1\Session): Colopl\Spanner\Session given.                                                                                   
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------- 
```
and when I removed ProtobufSpannerSession then error changed to 
```
 ------ --------------------------------------------------------------------------------------------------------------------------------- 
  Line   Colopl/Spanner/Concerns/ManagesSessionPool.php (in context of class Colopl\Spanner\Connection)                                   
 ------ --------------------------------------------------------------------------------------------------------------------------------- 
  85     Parameter #1 $protobufSession of class Colopl\Spanner\Session constructor expects Google\Cloud\Spanner\V1\Session, mixed given.  
 ------ --------------------------------------------------------------------------------------------------------------------------------- 

```

and according to https://phpstan.org/writing-php-code/narrowing-types
the only suggested way - explicit checking of types - via assert or directly

File: src/Colopl/Spanner/Concerns/ManagesStaleReads.php (hunk 3 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesStaleReads.php:76]: to mute phpstan error
```
 ------ ------------------------------------------------------------------------------------------------------------- 
  Line   Colopl/Spanner/Concerns/ManagesStaleReads.php (in context of class Colopl\Spanner\Connection)                
 ------ ------------------------------------------------------------------------------------------------------------- 
  73     Method Colopl\Spanner\Connection::selectOneWithTimestampBound() should return array|null but returns mixed.  
 ------ ------------------------------------------------------------------------------------------------------------- 
```

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 4 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php:167]: this is made specially to allow to run rollback without error
```
				} catch (Throwable $e) {
					$this->rollBack();
```
otherwise there is dangling transaction  and everything is breaking 
there is test for such use-case - testInTransactionSessionNotFoundHandledError

in more general sense - if session if lost, there are no ways to rollback transaction at all, the only manageable way - quietly abort current transaction and then rerun everything again

File: src/Colopl/Spanner/Connection.php (hunk 9 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: I force-pushed commit with fix

File: src/Colopl/Spanner/Connection.php (hunk 8 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: I force-pushed commit with fix

File: src/Colopl/Spanner/Connection.php (hunk 7 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: I force-pushed commit with fix

File: src/Colopl/Spanner/Connection.php (hunk 6 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: I force-pushed commit with fix

File: src/Colopl/Spanner/Connection.php (hunk 5 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: fixed in force-pushed pr

File: src/Colopl/Spanner/Concerns/ManagesStaleReads.php (hunk 2 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesStaleReads.php]: fixed in force-pushed pr

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php [hunk 10]
```diff
@@ -44,29 +45,31 @@ trait ManagesTransactions
...
+				} catch (Throwable $e) {
+					$this->rollBack();
+					throw $e;
+				}
+			}, ['maxRetries' => $attempts - 1]);
+
+			$this->fireConnectionEvent('committed');
+
+			return $return;
+		});
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Concerns/ManagesTransactions.php:72]: Please fix the tabs here to spaces as well.

File: src/Colopl/Spanner/Concerns/ManagesStaleReads.php (hunk 3 above)
Comment (taka-oyama) [reply src/Colopl/Spanner/Concerns/ManagesStaleReads.php:76]: Would adding `/** @var array<mixed>|null $result */` resolve the error as well?
I would like to avoid running asertions when we can get the same results without hindering performance.

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 4 above)
Comment (taka-oyama) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php:167]: Thank you. That makes sense. Could you add that last sentence as a comment?

File: src/Colopl/Spanner/Connection.php [hunk 11]
```diff
@@ -443,4 +462,83 @@ protected function runQueryCallback($query, $bindings, Closure $callback)
 
         return $result;
     }
+
+    /**
+     * Returns current mode
+     *
+     * @return string
+     */
+    protected function getSessionNotFoundMode()
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:471]:
```suggestion
    protected function getSessionNotFoundMode(): string
```

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php [hunk 12]
```diff
@@ -158,7 +161,14 @@ protected function performRollBack($toLevel)
 
         if ($this->currentTransaction !== null) {
             if ($this->currentTransaction->state() === Transaction::STATE_ACTIVE) {
-                $this->currentTransaction->rollBack();
+                try {
+                    $this->currentTransaction->rollBack();
+                } catch (NotFoundException $e) {
+                    // ignore session not found error
+                    if (empty($this->getSessionNotFoundMode()) || !$this->causedBySessionNotFound($e)) {
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Concerns/ManagesTransactions.php:168]: I think it's safe to remove `empty($this->getSessionNotFoundMode())` since it will always contain a value.

File: README.md [hunk 13]
```diff
@@ -236,6 +236,68 @@ In order to improve the performance of the first connection per request, we use
...
+[happen](https://cloud.google.com/spanner/docs/sessions#handle_deleted_sessions):
+
+ - Scripts that idle too long - for example, a [Laravel queue worker](https://laravel.com/docs/9.x/queues#the-queue-work-command)
+or anything that doesn't call Spanner frequently enough (more than once an hour).
+ - The session is more than 28 days old.
+ - Some random flukes on Google's side.
+
+The errors can be handled by one of the supported modes:
+
+- **MAINTAIN_SESSION_POOL** - When the 'session not found' error is encountered, the library tries to disconnect,
```
Comment (taka-oyama) [inline README.md:251]: I think the behavior for `CLEAR_SESSION_POOL` is good enough and this mode unnecessary.
By only having one way of doing things would make the whole thing simpler and easier to understand.
What do you think?

File: src/Colopl/Spanner/Concerns/ManagesSessionPool.php (hunk 1 above)
Comment (taka-oyama) [reply src/Colopl/Spanner/Concerns/ManagesSessionPool.php:85]: I'm starting to think phpstan level 9 might be too high for laravel based packages.
Would it bother you if I lowered it to 8? In which case I think this error will not appear here.

Comment (taka-oyama) [reply src/Colopl/Spanner/Concerns/ManagesSessionPool.php:85]: I just found out that asserts can be tuned to not have performance impact on production so this code is OK.

https://www.php.net/manual/en/function.assert.php#function.assert.expectations

File: src/Colopl/Spanner/Concerns/ManagesStaleReads.php (hunk 3 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesStaleReads.php:76]: This is why I used assert(), it has special feature in recent php - any code inside it doesn't run in production mode
from https://www.php.net/manual/en/ini.core.php#ini.zend.assertions
>When set to -1, assertion code will not be generated, making the assertions zero-cost (production mode).
but I will do as you prefer

Comment (taka-oyama) [reply src/Colopl/Spanner/Concerns/ManagesStaleReads.php:76]: Yeah. I just read up on this. Sorry for the confusion. This can stay as is.

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 4 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php:167]: sure, also I found edge case when exception anyway should be called - i.e. on explicit rollback inside transaction({}), 
I am adding test-case and code for handling

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 12 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php:168]: I thought to allow disabling mode by simple 
```
   	        'sessionNotFoundErrorMode' => false,
```
I am going to redo it to 
   	        'sessionNotFoundErrorMode' => 'THROW_EXCEPTION',

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 10 above)
Comment (taka-oyama) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php:72]: Sorry for the ambiguity. The entire fix here is indented with tabs. Please replace the tabs with 4 spaces.

File: README.md (hunk 13 above)
Comment (oprudkyi) [reply README.md:251]: Clearing session pool (while really is needed in very rare occasions)  is very 'hard' operation, which implicitly affects all processes that use session poll, i.e. for example if there 50 (httpd/php-fpm) workers that share session pool, then after clearing they all simultaneously will receive 'session not found' (because clearing deletes sessions) and end with restarting potentially heavy and long running scripts
so MAINTAIN_SESSION_POOL is the most safe mode - it still handles most of the cases (like idle workers), but do not introduce hard reset of sessions

so there are possible options
- left MAINTAIN_SESSION_POOL as safest way for users who doesn't want running into potentially risking hard reset
- left it and make it default 
- remove it, but provide users some level of control 
```
function setClearSessionPoolOnSessionNotFoundError(bool $value)
```
though it seems ugly and less clean

File: src/Colopl/Spanner/Connection.php [hunk 14]
```diff
@@ -443,4 +462,83 @@ protected function runQueryCallback($query, $bindings, Closure $callback)
...
+
+    /**
+     * Handle "session not found" errors
+     *
+     * @template T
+     * @param  Closure(): T $callback
+     * @return T
+     * @throws InvalidArgumentException|NotFoundException|AbortedException
+     */
+    protected function sessionNotFoundWrapper(Closure $callback)
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:484]: Can you rename this to something like
```suggestion
    protected function withSessionHandling(Closure $callback): mixed
```

File: README.md (hunk 13 above)
Comment (taka-oyama) [reply README.md:251]: I understood your concern. I now believe this can be as is.

However, now I'm thinking the opposite. Is `CLEAR_SESSION_POOL` necessary?

Wouldn't each session be retried eventually anyway?

Comment (oprudkyi) [reply README.md:251]: > Is CLEAR_SESSION_POOL necessary
yes, it is necessary at least in my use case - without it there is error once a week or two and maintain() didn't help

Comment (taka-oyama) [reply README.md:251]: So I'm guessing the error still occurred because in a case where `MAINTAIN_SESSION_POOL` is used, the handler only handles the error once, and upon reconnection, the logic probably choose another session that was also expired and errored out.
Is my understanding correct?

Comment (oprudkyi) [reply README.md:251]: no, There are rare cases when google spanner deletes a session entirely, while the session is still accounted in session pool as active, so maintain and reconnection just reload it again and again. and the only way to fix it is deletion of the cache   file manually (or via code, or restarting docker/pod)

File: src/Colopl/Spanner/Connection.php (hunk 14 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: fixed in recent commit

File: README.md (hunk 13 above)
Comment (oprudkyi) [reply README.md]: `MAINTAIN_SESSION_POOL` set as default mode

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 12 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php]: done in recent commits

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 10 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php]: done

File: src/Colopl/Spanner/Concerns/ManagesTransactions.php (hunk 4 above)
Comment (oprudkyi) [reply src/Colopl/Spanner/Concerns/ManagesTransactions.php]: done, please check

File: src/Colopl/Spanner/Connection.php [hunk 15]
```diff
@@ -443,4 +467,85 @@ protected function runQueryCallback($query, $bindings, Closure $callback)
...
+        }
+
+        try {
+            return $callback();
+        } catch (NotFoundException $e) {
+            // ensure if this really error with session
+            if ($this->causedBySessionNotFound($e)) {
+                if ($this->inTransaction()) {
+                    // if we inside transaction then throw abort exception
+                    throw new AbortedException(self::SESSION_NOT_FOUND_CONDITION, $e->getCode(), $e);
```
Comment (taka-oyama) [inline src/Colopl/Spanner/Connection.php:513]: I think we should throw this error after reconnecting.
That way we don't have to catch the AbortedException in the transaction code.

Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php:513]: Hi @taka-oyama 
this won't help - [Transaction](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/Transaction.php) is tied to a session and stores it internally in private property. 
I tried to de-duplicate code in catch, please check 
https://github.com/colopl/laravel-spanner/pull/39/commits/df149c64df5c448f58c0e767b96465c2524c85a5

Comment (taka-oyama) [reply src/Colopl/Spanner/Connection.php:513]: So In that case, wouldn't it make sense to just throw the NotFoundException?

Wouldn't throwing AbortedException here [cause the transaction to retry](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/Database.php#L864) with the same session?

Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php:513]: >So In that case, wouldn't it make sense to just throw the NotFoundException?
no, this will not work with reentrancy of withSessionNotFoundHandling

>Wouldn't throwing AbortedException here [cause the transaction to retry](https://github.com/googleapis/google-cloud-php/blob/main/Spanner/src/Database.php#L864) with the same session?

yes, I am going to reimplement with special exception class , something like TransactionRestartException

Comment (oprudkyi) [reply src/Colopl/Spanner/Connection.php]: Hi @taka-oyama 
I have implemented it without extra exception - just by handling ordinal QueryException

Comment (taka-oyama) [reply src/Colopl/Spanner/Connection.php]: Thank you for the fix!

**Comment by taka-oyama** [comment]
Thank you for your contribution!

**Comment by taka-oyama** [comment]
I will release the next version once I fix the Phpstan errors.

**Comment by oprudkyi** [comment]
Hi @taka-oyama 

Thank you. 
could you also please create branch v3 from https://github.com/colopl/laravel-spanner/tree/v3.9.3
so it will be possible to backport to older laravel ?

**Comment by taka-oyama** [comment]
Hi, thanks for the patience.

Yes, I will backport it.

I couldn't find time this week so I will do it on monday.

**Comment by oprudkyi** [comment]
HI @taka-oyama 
I thought to make a backport by self - the my whole endeavor was to get universal solution for Laravel 8.
But any solution will do.

**Comment by taka-oyama** [comment]
Hi. I released v4.1.0 and 3.10.0 (for laravel 8 or below), please check it out.

**Comment by taka-oyama** [comment]
[Asking google to fix this upstream](https://github.com/googleapis/google-cloud-php/issues/5827), so we don't have to do this hack.
